oap-server/server-receiver-plugin/skywalking-meter-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/meter/provider/handler/BatchMeterServiceHandler.java

```diff
@@ -0,0 +1,92 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.receiver.meter.provider.handler;
+
+import io.grpc.stub.StreamObserver;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.skywalking.apm.network.common.v3.Commands;
+import org.apache.skywalking.apm.network.language.agent.v3.MeterData;
+import org.apache.skywalking.apm.network.language.agent.v3.MeterDataCollection;
+import org.apache.skywalking.apm.network.language.agent.v3.MeterReportServiceGrpc;
+import org.apache.skywalking.oap.server.analyzer.provider.meter.process.IMeterProcessService;
+import org.apache.skywalking.oap.server.analyzer.provider.meter.process.MeterProcessor;
+import org.apache.skywalking.oap.server.library.module.ModuleManager;
+import org.apache.skywalking.oap.server.library.server.grpc.GRPCHandler;
+import org.apache.skywalking.oap.server.telemetry.TelemetryModule;
+import org.apache.skywalking.oap.server.telemetry.api.CounterMetrics;
+import org.apache.skywalking.oap.server.telemetry.api.HistogramMetrics;
+import org.apache.skywalking.oap.server.telemetry.api.MetricsCreator;
+import org.apache.skywalking.oap.server.telemetry.api.MetricsTag;
+
+/**
+ * Meter protocol receiver, collect and process the meters.
+ */
+@Slf4j
+public class BatchMeterServiceHandler extends MeterReportServiceGrpc.MeterReportServiceImplBase implements GRPCHandler {
+
+    private final IMeterProcessService processService;
+    private final HistogramMetrics histogram;
+    private final CounterMetrics errorCounter;
+
+    public BatchMeterServiceHandler(ModuleManager manager, IMeterProcessService processService) {
+        this.processService = processService;
+        MetricsCreator metricsCreator = manager.find(TelemetryModule.NAME)
+                                               .provider()
+                                               .getService(MetricsCreator.class);
+        histogram = metricsCreator.createHistogramMetric(
+            "meter_in_latency", "The process latency of meter",
+            new MetricsTag.Keys("protocol"), new MetricsTag.Values("grpc")
+        );
+        errorCounter = metricsCreator.createCounter("meter_analysis_error_count", "The error number of meter analysis",
+                                                    new MetricsTag.Keys("protocol"),
+                                                    new MetricsTag.Values("grpc")
+        );
+    }
+
+    @Override
+    public StreamObserver<MeterDataCollection> collectBatch(StreamObserver<Commands> responseObserver) {
+        return new StreamObserver<MeterDataCollection>() {
+            @Override
+            public void onNext(MeterDataCollection meterDataCollection) {
+                final MeterProcessor processor = processService.createProcessor();
```

Comment:
I think this should not be created inside `onNext`. Take a close look at `MeterServiceHandler`. It is created once for a streaming only.

Reply:
MeterProcessor stores the data of a matricCollection. If we want to reuse, we must provide the clean method in MeterProcessor. It does not support reuse of this data structure.